tests/integrate_test/iiss/prevote/test_prep.py

```diff
@@ -314,145 +314,179 @@ def test_reg_prep_validator(self):
+        self._validate_name(reg_data)
+        reg_data = deepcopy(prep_register_data)
+        self._validate_email(reg_data)
+        reg_data = deepcopy(prep_register_data)
+        self._validate_website(reg_data)
+        reg_data = deepcopy(prep_register_data)
+        self._validate_country(reg_data)
+        reg_data = deepcopy(prep_register_data)
+        self._validate_city(reg_data)
+        reg_data = deepcopy(prep_register_data)
+        self._validate_details(reg_data)
+        reg_data = deepcopy(prep_register_data)
+        self._validate_p2p_endpoint(reg_data, revision)
 
     def test_reg_prep_validator_fixed_email_validation(self):
         self.update_governance()
 
         # set Revision REV_FIX_EMAIL_REGEX
-        self.set_revision(Revision.FIX_EMAIL_VALIDATION.value)
+        revision = Revision.FIX_EMAIL_VALIDATION.value
+        self.set_revision(revision)
 
         # gain 10 icx user0
         balance: int = 10 * ICX_IN_LOOP
         self.distribute_icx(accounts=self._accounts[:8],
                             init_balance=balance)
 
-        self._validate_name()
-        self._validate_fixed_email()
-        self._validate_website()
-        self._validate_country()
-        self._validate_city()
-        self._validate_details()
-        self._validate_p2p_endpoint()
+        reg_data = deepcopy(prep_register_data)
+        self._validate_name(reg_data)
+        reg_data = self.get_prep_data_with_customized_endpoint("a")
+        reg_data[ConstantKeys.EMAIL] = '你好@validexample.com'
+        self._validate_fixed_email(reg_data)
+        reg_data = self.get_prep_data_with_customized_endpoint("b")
+        self._validate_website(reg_data)
+        reg_data = self.get_prep_data_with_customized_endpoint("c")
+        self._validate_country(reg_data)
+        reg_data = self.get_prep_data_with_customized_endpoint("d")
+        self._validate_city(reg_data)
+        reg_data = self.get_prep_data_with_customized_endpoint("e")
+        self._validate_details(reg_data)
+        reg_data = self.get_prep_data_with_customized_endpoint("f")
+        self._validate_p2p_endpoint(reg_data, revision)
+
+    def test_prevent_duplicated_endpoint(self):
+        self.update_governance()
+
+        # set Revision REV_FIX_EMAIL_REGEX
+        revision = Revision.PREVENT_DUPLICATED_ENDPOINT.value
+        self.set_revision(revision)
+
+        # gain 10 icx user0
+        balance: int = 10 * ICX_IN_LOOP
+        self.distribute_icx(accounts=self._accounts[:2],
+                            init_balance=balance)
+
+        reg_data = deepcopy(prep_register_data)
+        tx = self.create_register_prep_tx(self._accounts[0], reg_data)
+        self.process_confirm_block_tx([tx])
+
+        # fail to register due to duplicated endpoint
+        tx = self.create_register_prep_tx(self._accounts[1], reg_data)
+        self.process_confirm_block_tx([tx], expected_status=False)
+
+        # success to set prep
```

Comment:
Please add a failure case for testing setPRep() with duplicated p2p_endpoint of which format is  IP:PORT.

Reply:
Please use "1.2.3.4:7100" instead of node2.endpoint:7100.
You will find a bug in _validate_p2p_endpoint()